### ISSUE TITLE
error for predict with newdata if splines/poly with different df/degree in fixed and random effects

### ISSUE THREAD

**Issue by rdiaz02** [glmmTMB/glmmTMB#632]
If we fit models with splines (say, `ns`) with different df for a term in the fixed and random effects, `predict` gives an error if we pass `newdata`. Same for polynomials with different degree for a term in fixed and random effects. For example:

```
library(glmmTMB)
library(splines)
data(sleepstudy,package="lme4")
m4 <- glmmTMB(Reaction ~ ns(Days, df = 3) + (ns(Days, df = 2)|Subject), 
              data = sleepstudy)
predict(m4, newdata = data.frame(Days = 4, Subject = "372"), re.form = NULL, 
          type = "response")
```

gives error

```
Error in value[[jj]][ri, ] <- xij : 
  number of items to replace is not a multiple of replacement length
```

These two also fail
```
p4 <- glmmTMB(Reaction ~ poly(Days, degree = 3) + (poly(Days, degree = 2)|Subject), 
              data = sleepstudy)
predict(p4, newdata = data.frame(Days = 4, Subject = "372"), re.form = NULL, type = "response")

p5 <- glmmTMB(Reaction ~ poly(Days, degree = 2) + (poly(Days, degree = 1)|Subject), 
              data = sleepstudy)
predict(p5, newdata = data.frame(Days = 4, Subject = "372"), re.form = NULL, type = "response")
```

Note that this (and also for the other fitted models above)
```
predict(m4,  re.form = NULL, type = "response")
```
is OK. 

And this, equivalent to `p5`, but without using `poly` in the random effect specification, also works:
```
p6 <- glmmTMB(Reaction ~ poly(Days, degree = 2) + (Days|Subject), 
              data = sleepstudy)
predict(p6, newdata = data.frame(Days = 4, Subject = "372"), re.form = NULL, type = "response")
```

It is also OK to obtain predictions using `newdata` from models such as

```
m1 <- glmmTMB(Reaction ~ ns(Days, df = 3) + (Days|Subject), 
              data = sleepstudy)
m2 <- glmmTMB(Reaction ~ Days + (ns(Days, df = 3)|Subject), 
              data = sleepstudy)
m3 <- glmmTMB(Reaction ~ ns(Days, df = 2) + (ns(Days, df = 2)|Subject), 
              data = sleepstudy)
predict(m1, newdata = data.frame(Days = 4, Subject = "372"), re.form = NULL, type = "response")
predict(m2, newdata = data.frame(Days = 4, Subject = "372"), re.form = NULL, type = "response")
predict(m3, newdata = data.frame(Days = 4, Subject = "372"), re.form = NULL, type = "response")
```

**Comment by bbolker** [comment]
I was kind of afraid of this.  This sort of thing is usually a bit of a pain to sort out.  I agree that it should work, but I'm curious what your use case is ... ??

**Comment by rdiaz02** [comment]
Sorry (we can always pretend this never happened :-) ).

Use case: briefly, to plot data with lines for the fit I'd normally do, say, `geom_lines` with a large number of predicted values (via `newdata`). Why? Because if some individuals have only a few observations, using `predict` without `newdata` will give only a few predictions, and the lines for those individuals, instead of looking like nice curves are, well, a set of seven or eight connected straight segments. (So, yes, I guess I would agree that this is basically a aesthetic issue, and as far as my searching around shows, no one has had a similar case).

More details: the models I am fitting are for a beta model (variable in (0, 1) ) using things such as `ns(x, 3)`  in the fixed effects but only `ns(x, 2)`. Why the difference in df? Because using `ns(x, 3)` in the random effects leads to convergence issues, I think related to some individuals having few observations.

**Comment by bbolker** [comment]
Well, we should fix this in any case (although at least an error is better than incorrect output!). I completely understand why you want to use `newdata` ... the part that's weird is using different `ns()` complexities in the fixed and random components - it's mathematically fine, but odd in terms of modeling.  A more principled approach would be to try to penalize the random effects covariance matrix to improve convergence. Not necessarily easy, though. As a side thought, could you make your model work in `mgcv` ?

**Comment by rdiaz02** [comment]
> (although at least an error is better than incorrect output!). 

Yes, fully agree: incorrect output would have been a lot worse (as of now, it just a minor inconvenience for plots).

> the part that's weird is using different `ns()` complexities in the fixed and random components - it's mathematically fine, but odd in terms of modeling. 

Yes, I still feel somewhat uncomfortable. This is my brain dump (apologies if it is obvious or out of place):

To try to grasp the thing intuitively, I reasoned that it is very common to have models with a fixed effects regression with a random intercept; and we say something like "parallel lines, shifted up or down", and this we find easy to understand. 

What if we model the fixed effects with a degree 2 polynomial and the random effects with a degree 1 (straight line)? Now we add curvature to the fixed effect line while the random effects only affect the linear term (in addition to the intercept); so the random effects affect the linear rate of change, not the curvature. 

And then we repeat once more: cubic for fixed effects, only up to quadratic for random. I see the logic, but I don't fully have a graphical intuition (I spent an hour plotting parabollas and overimposing "random effects" parabollas with random deviations in the different terms ---intercept, linear, quadratic-- to try to "really get it", and I still find it non-obvious).

In the end, since there were constraints due to convergence (which I think might have to do with correlations between the random effects terms) and since it is mathematically fine, I thought this was a lot more reasonable than only adding, say, a random effects intercept.

> A more principled approach would be to try to penalize the random effects covariance matrix to improve convergence. Not necessarily easy, though. 

I only tried forcing the different random effects parts to be uncorrelated (the regressor has a natural intercept at 0, as it is a proportion). It did not help.

> As a side thought, could you make your model work in `mgcv` ?

Thanks for the suggestion. I just tried, with packages `mgcv` (for function `gamm` ) and `gamm4` (function `gamm4`). But I think this will not work in my case, since to model a beta regression I need to use, I believe, `family = betar(link = "logit")` (https://stat.ethz.ch/R-manual/R-patched/library/mgcv/html/Beta.html) which is an "extended family".  `gamm` tells me `gamm is not designed to use extended families` and `gamm4` gives a similar message (` internal problem in as(): “extended.family” is(object, "family") is TRUE, but the metadata asserts that the 'is' relation is FALSE`)

**Comment by bbolker** [comment]
fixed by f0765ba9bd ?

**Comment by bbolker** [comment]
@rdiaz02 Can you comment on whether this is fixed in https://github.com/glmmTMB/glmmTMB/pull/854 ?

**Comment by rdiaz02** [comment]
Sorry, I've just seen this. Swamped right now with teaching and a couple of deadlines for mid october. I should be able to look at it in a couple of weeks.
